### ISSUE TITLE
Update and validate build for windows

### ISSUE THREAD

**Issue by ColinPitrat** [ColinPitrat/caprice32#11]
Currently, the build for windows is not maintained and is probably not working anymore.
The windows makefile should be removed and the build for windows should be done from the main makefile using cross-compilation.
Validation for it should be added on travis

**Comment by ColinPitrat** [comment]
`make windows` is now cross-compiling for windows and copying all that should be needed in a win subdirectory.

However, the build is failing because of the usage of `realpath` function which is not available on windows.

**Comment by ColinPitrat** [comment]
Build for windows is now successful but running cap32.exe fails on wine with:
`terminate called after throwing an instance of 'std::logic_error'
what(): basic_string::_M_construct null not valid`

It runs correctly (except for some sounds bugs) on my wife's laptop on Windows 7 but the issue above is reported on native Windows 10 (cf issue #54).

There seems to be something odd with the result of the cross-compilation. This may be link with the fact that it depends on msvcrt.dll that is not delivered and differ from one system to another.

**Comment by sebhz** [comment]
See below comment copied from issue #54 (duplicate of this one).
Since I do happen to have access to a couple of MS Windows machines, I can try and handle "native" compilation under MSYS/mingw-64 under windows. Of course this strange problem with GUI options freezing the emulator has to be solved also :-)

==================================================================
For what it is worth... I did spend some time today to compile Caprice32 on a Windows 8 machine using MSYS2/mingw64. Once all the dependencies are installed compilation is pretty straightforward. cap32 runs fine with it - I tried two disks (sorcery and trailblazer), and I get 100 fps consistently. Audio seems correct. I'll try the same binary on my wife's laptop (Win10) tonight. If it works fine, I'll share it !

Note that there is one problem with this executable: trying to invoke the "Option" menu freezes the emulator completely. This is painful but you can still use the emulator by changing manually the settings in the .cfg file. I'll look into this.

**Comment by sebhz** [comment]
Option menu freeze -> stuck in an endless loop of "wGui exception while delivering message : Unable to load bitmap" :-)

**Comment by ColinPitrat** [comment]
Are there some missing files in resources ?

**Comment by sebhz** [comment]
Yep - just found this. I'll submit a patch in a few minutes for it.
And open an issue - we should not freeze if this is the case :-)

**Comment by ColinPitrat** [comment]
OK, I'm investigating the crash (to avoid double work).

**Comment by sebhz** [comment]
Done - see my patch :-)

**Comment by ColinPitrat** [comment]
For info, my patch: https://github.com/ColinPitrat/caprice32/commit/f110b53f354e0d55fe3fd7aa1d2ee106cca0e486

**Comment by sebhz** [comment]
Some more (bad) news with this. I am actually getting the same error when running the application outside of MSYS2 shell.

"terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information."

I'll keep investigating this. We should have proper MS Windows support.

**Comment by ColinPitrat** [comment]
Do you mean it works well under msys2 shell and fails outside of it ?

**Comment by sebhz** [comment]
Yes... I thought about the msvcrt.dll trick, but interestingly the is no msvcrt.dll under the whole msys2 / mingw-64 file system. There is a libmsvcrt.a under /lib, though...

**Comment by ColinPitrat** [comment]
OK, same as my mingw setup then ... My plan for this evening was to try with a simple hello world to see what it gives.

**Comment by sebhz** [comment]
Trying to understand the error message, would it be possible that we actually initialize a std::string with a NULL pointer somewhere in the caprice32 code (though unwittingly), and that causes the issue.
Or rather, for some reason gcc compiles the code properly (doing typecasting under the carpet without us seeing it) but not mingw-64. This does not look like a runtime issue.

**Comment by ColinPitrat** [comment]
Yes, that's typically the message you would have by constructing a string with a null pointer.
However if you try to add logs at the very beginning of main, they won't even get printed.
Or if you try to run in gdb and use 'start' (which basically place you at main entrance) it fails before.

So it looks like static initialization issue. This could very well be linked to the order of their initialization in fact.

**Comment by ColinPitrat** [comment]
Good news is there are only 254 global variables in caprice (we definitely need to clean this !):
cpitrat@cpitrat-l:~/Perso/caprice32$ objdump -t cap32 | grep " g " | grep -v ".text" | wc -l
254

**Comment by sebhz** [comment]
Does not make sense. Why on Earth would it work under the msys2 or mingw64 shell then, if it is a static init issue ? I did cross-compile using mingw-64, same as you...

**Comment by ColinPitrat** [comment]
I think the static initialization fiasco could explain it:

**Comment by ColinPitrat** [comment]
I did a few tests this evening and it appears that removing the GUI fixes the issue.
This narrows down the source of the problem to the GUI code and confirms this is an issue with caprice code.

**Comment by ColinPitrat** [comment]
I found the culprit in src/gui/includes/wgui_include_config.h:
`static const std::string USER_CONFIG_PATH = std::string(getenv("HOME")) + "/.wgui/";`

If getenv returns NULL, this fails. This header (and two of his friends) is totally useless. I'm cleaning this up.

**Comment by sebhz** [comment]
Nice catch !

I made a build with the latest head on my Win8 machine under MSYS2. In order for it to run natively on windows, I had to embed the following 6 extra DLLs: libglib-2.0-0, libgraphite2, libharfbuzz-0, libiconv-2, libintl-8, libpcre-1. I did not investigate why those would be needed (glib2 ?!), but this is the minimum set I need. Is it an MSYS2 thing, or do you see the same on your end under Wine ?

Once this is done, the executable works fine. I tried Win8 and Win10. No problem. Dragging and dropping a dsk image on the executable icon does what it is supposed to do. I am getting capped at 50 fps consistently. Sound emulation on my two preferred games (sorcery and trailblazer) seems OK.

**Comment by sebhz** [comment]
And BTW... with a small modification to the makefile (HAVE_GL added to WINCFLAGS), compilation with HAVE_GL works fine, and we can get a nice full screen OpenGL-rendered UI with scanlines under MS Windows too.

**Comment by ColinPitrat** [comment]
Right. This was a miss, which is not surprising considering the duplication between linux and windows builds.
I'm trying to refactor the Makefile so that the same rules are used to build for win32, win64 and linux. This way the same flags should work for all platforms.

**Comment by ColinPitrat** [comment]
Windows and Linux builds are now using the same rules and the differences are reduced to a few variables having different values depending on ARCH value (linux, win32 or win64). It's possible to run unit_test with windows build (using wine).

Some improvements would still be nice:
 - the Windows build is not perfect with paths hard coded for my ArchLinux environment.
 - the platform specific parts could be extracted in specific files that would be imported in the makefile (this would avoid some duplication, for example to list all the existing options for ARCH)
 - it seems there are some issues with sound in windows build that do not appear on linux (e.g in Fruity Frank, the sound when a purple monster appears)
 - there is no automatic validation of windows build & tests. Setting this up on travis doesn't seem easy because of the dependencies that are not available as packages for mingw.

Help for any of those point (especially the last one) is welcome :-)

**Comment by sebhz** [comment]
The first point is hardly a problem, at least for me. I am compiling under MS Windows MSYS2, using 'make ARCH=win64 MINGW_PATH=/mingw64 and this works like a breeze. Well, it crashes when trying to compile debug builds since wine is obviously not installed on my setup, but this is hardly a big problem, and probably easy to fix. I'll submit a patch for this in the makefile, and update the INSTALL.md instructions to describe MSYS2 build dependencies.

I'll have a look at the sound thing. Fruity Frank has always be one of my fave games anyway :-)

**Comment by ColinPitrat** [comment]
If making the build work under windows is relatively easy, maybe we could try AppVeyor for the last point.

**Comment by ColinPitrat** [comment]
After 32 iterations, we now have continuous build & test for Windows !
https://ci.appveyor.com/project/ColinPitrat/caprice32

**Comment by ColinPitrat** [comment]
Keeping the issue opened for the two last points:
- the platform specific parts could be extracted in specific files that would be imported in the makefile (this would avoid some duplication, for example to list all the existing options for ARCH)
- it seems there are some issues with sound in windows build that do not appear on linux (e.g in Fruity Frank, the sound when a purple monster appears)

**Comment by sebhz** [comment]
* On my Windows PC, the violet monster sound in Fruity Frank seems perfect to me.
* Regarding architecture specific makefile parts, I'd rather keep on single file even if it is a bit convoluted than multiple ones. It looks clearer to me - but it's a metter of personal taste. Your call anyway.
* Oh, and of course, now that you added IPF support on Linux, we need it on Windows too :-) I had a look at sps developer distro and access API for windows (the user distrib only includes a DLL), and their headers are not compatible with the Linux one. Looks like a significant piece of work would be needed here. Anyway you are mostly on your own there since I do not have any IPF file to play with, and they seem pretty hard to find on the Internet.

**Comment by ColinPitrat** [comment]
For the IPF files, there are some available here: http://www.cpc-power.com/index.php?page=database&ipf=1
I didn't notice the header incompatibility ! I thought there would be no code change needed and just using the DLL would work (the code was initially written for windows and I had little adaptations to do). Maybe this is just a matter of header namings and organization and linking with the DLL would actually work ... 

Then there's the problem of distribution of the resulting binary as the license of SPS is a special one. As far as I understand it should be OK as long as it's free. I don't want to integrate it into the repository as I want to keep any usage of caprice compatible with the GPL v2 straightforward but I think distributing an archive for a windows build including the library should be OK.

Anyway, I submitted the code by mistake, I wanted to only commit it locally for now (I'm so used to push after commit that it became an habit :-)). I'm waiting for the confirmation of the original author that he's OK for his code to be placed under GPL v2.

For the architecture specifics extraction, I'll give it a try to see what it gives. It may or may not bee a good idea indeed, I suppose there shouldn't be much more architectures in the future :-)

**Comment by sebhz** [comment]
Forget what I said about IPF under windows. I could make it compile and link without problem under Windows. Will submit a patch shortly. I could still not check if this worked fine, since the IPF files are not available anymore on cpc-power, it seems.

**Comment by ColinPitrat** [comment]
If you contact me by mail at first_name.last_name on gmail I can send you some images so that you can test.

**Comment by ColinPitrat** [comment]
And we now have a "Caprice32 latest" release that, unless I made a mistake, is automatically updated at each push with two newly built archives for Windows (32 & 64) which is pretty cool.

**Comment by sebhz** [comment]
Actually, if you want the release to be usable as is on an MS Windows machine and without MSYS2, you are missing a few libraries.
Funnily enough under MSYS2, the freetype library depends on libglib-2.0-0, libgraphite2, libharfbuzz-0, libiconv-2, libintl-8, libpcre-1 (you can see this from the Appveyor log, when freetype is installed). The releases as they are today are not usable outside of the MSYS2/mingw64 environment - I did a quick trial on my Windows machine two minutes ago, caprice32 would not start. I will provide a patch to address this, but I am afraid it might break your archlinux cross-compilation environment.

**Comment by ColinPitrat** [comment]
If we have working build on appveyor, it's not an issue to break mine :-)
An option though is to do something like:
`cp $(MINGW_PATH)/bin/libglib-2.0-0.dll $(ARCHIVE)/ || echo "libglib-2.0-0.dll not found - ignoring"`
Something like that.

**Comment by ColinPitrat** [comment]
In order to not forget it, three more things to do + recap of previous remaining TODO:
 - add support for IPF in windows build on AppVeyor (validate build with & without and deliver with)
 - support creating a release in both appveyor and travis (or is only appveyor enough ?) and release Caprice 4.4
 - add some "end to end tests" (start caprice, execute some commands, force a print screen (needs to be implemented) and check output)
 - see if there's some benefit in extracting the platform specific parts from the makefile (this would avoid some duplication, for example to list all the existing options for ARCH)
 - retry Fruity Frank to clarify the issue I observed with sound when a purple monster appears

**Comment by sebhz** [comment]
Is IPF support "GPL v2.0" clean ?

**Comment by ColinPitrat** [comment]
For the code to support it in caprice32, I got the confirmation that we can include it.
For redistributing the library itself, we need to distribute the license with it. This will be the opportunity to also include the other licenses that are in the licenses subdirectory (for the fonts and for wGui).

**Comment by ColinPitrat** [comment]
The 3 first points are done.

For the makefile, I'll let it like that for now and we'll see in the future if more evolutions are needed.

On Fruity Frank side, the sound definitely behaves strangely on my wife windows. The sound when eating a fruit is monotonic and doesn't stop until another sound (other than the music, usually the purple monster falling) occurs. The sound of the purple monster is also monotonic but does stop when it reaches the ground. Anyway if you don't have the same issue it means it's not a generic windows problem. I'll open a separate bug to track it.

**Comment by zackizacki** [comment]
Are there some todo's left, or can this be closed?

**Comment by ColinPitrat** [comment]
Looking at the bug, I think we should do the following:
 - add IPF licence in the licenses subdirectory
 - distribute the licenses directory in the releases

**Comment by ColinPitrat** [comment]
Actually, the license is already copied in the windows build.
For the sources, no need to include the license as we don't have the library in the repository and the user has to download it themselves.

**Comment by ColinPitrat** [comment]
Note that this issue is already closed since Mar 2017 :-)
